build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -586,6 +590,21 @@ message DirectoryNode {
   Digest digest = 2;
 }
 
+// A `SymlinkNode` represents a symbolic link.
+message SymlinkNode {
+  // The name of the symlink.
+  string name = 1;
+
+  // The target path of the symlink. The path separator is a forward slash `/`.
+  // The target path can be relative to the parent directory of the symlink or
+  // it can be an absolute path starting with `/`. Absolute paths are
+  // interpreted relative to the input root directory, i.e., they are not
+  // allowed to escape the input space. The canonical form forbids the
+  // substrings `/./` and `//` in the target path. `..` components are allowed
```

Comment:
Can we restrict `..` to only be allowed to be leading components, and only for relative symlinks?

I'm looking to avoid two things:
1. `foo/../bar` which is redundant and non-canonical - this feels trivial for callers to avoid, so should be forbidden
2. `/../foo` which is not allowed because it would escape the input root directory - explicitly saying both "they are not allowed to escape the input space" and "`..` components are allowed anywhere" seems to open up an ambiguity that we can avoid.

The text I would suggest is "The canonical form forbids the substrings `/./` and `//` in the target path. `..` components are allowed only in relative paths, and only as leading components in the target path, i.e. any substring `../` may only be preceded by 0 or more `../` literals."

Reply:
> 2. `/../foo` which is not allowed because it would escape the input root directory - explicitly saying both "they are not allowed to escape the input space" and ".. components are allowed anywhere" seems to open up an ambiguity that we can avoid.

Following POSIX semantics, `/../foo` would be identical to `/foo` and thus also not escape the input root directory. However, as this is trivial to canonicalize, I agree that we should not allow symlink targets starting with `/../`.

> 1. `foo/../bar` which is redundant and non-canonical - this feels trivial for callers to avoid, so should be forbidden

This is not as simple as it may seem. If `foo` is a symlink to a directory, `foo/../bar` is not guaranteed to be equivalent to `bar` (`..` must be resolved after following `foo`). And `foo` might even be a dangling symlink that becomes valid as part of the build process. To be on the safe side with regards to compatibility I prefer allowing `..` components in symlink targets.

Reply:
OK, I re-read the whole discussion in the document comments about absolute paths now, and recall the problems. I'm still not sure whether to allow absolute paths in general, and only block them on a per-platform or per-server basis. I'm not a big fan of treating absolute paths as relative to the input root, because that is very confusing, and I'm not sure what problem will it solve -- packages that create absolute symlinks would still need to be heavily amended, right?
Suggestion: how about we simply disallow absolute paths *for now*, and will add them later (it is a non-breaking change) when particular use-cases arise?

Reply:
> packages that create absolute symlinks would still need to be heavily amended, right?

BuildStream will use workers that build in sandboxes where the input root is the filesystem root and thus, packages can use absolute symlinks without any modifications. I understand that this doesn't apply to other platforms, which is why we propose to leave this decision to each platform.

If this is not currently acceptable, we could definitely follow your suggestion and disallow absolute paths for now and add them later after more discussion.

Reply:
Bleh. Okay, how about the server returning something like:

    // Describes how the server treats absolute symlink targets.
    enum SymlinkAbsolutePathStrategy {
      UNKNOWN = 0;

      // Server will return an INVALID_ARGUMENT on symlinks with absolute targets.
      DISALLOWED = 1;

      // Server will allow symlink targets to escape the input root tree, possibly resulting in
      // non-hermetic builds.
      ALLOWED = 2;

      // Server will treat absolute symlink targets as relative to the input root, i.e. for "/a/foo"
      // an absolute symlink "/worker123/.../input_rootXYZ/a/foo" will be created.
      INPUT_ROOT = 3;
    }
    message CacheCapabilities {
       ...
       SymlinkAbsolutePathStrategy symlink_absolute_path_strategy = 5;
    }

And then you can just refer to that enum in the comments.

Reply:
Although, wait, that affects the client behavior as well, for symlinks returned from the server. In the ALLOWED case, the client should just symlink the exact same absolute target non-hermetically; for INPUT_ROOT, the client should do the same translation? But bleh, that is just so non-intuitive...

Wait, we already allow setting the working directory in the API. What if we restrict the enum to just allow/disallow for now, and BuildStream just sets the working directory to "/" for its actions? In that case, allowing absolute paths is the same as treating them as input root, right?

Reply:
I'm going to chime in and I think that ALLOWED and DISALLOWED should be the only behaviours allowed; INPUT_ROOT I think has too much ambiguity to it. If we require the server to reinterpret symlinks in any way, the behaviour could end up depending on whether the server creates the symlink as an absolute or relative path (because the symlink `/input_root/foo` and `../foo` are semantically different, even if they both appear in `/input_root/bar`, and if `/input_root` is a symlink to `/real_root`, is the server expected to make a link to `/real_root/foo`?). I worry that trying to specify exactly how they should work is going to get confusing especially on Windows where [path handling is very complicated](https://googleprojectzero.blogspot.com/2016/02/the-definitive-guide-on-win32-to-nt.html).

I see concerns about symlinks escaping the input root, but I can't say that I really understand them. There's plenty of other ways that commands can escape the input root, and I don't think a command that runs `/usr/bin/gcc` is any less dependent on the worker environment than one that runs `./gcc` which is a symlink to `/usr/bin/gcc`, say.

As for canonicalization, realistically, it is more strongly a set of best practices than hard-and-fast rules. The advantage to encoding them into the API, rather than letting the client do what is best, is merely to let the server detect and reject non-canonical actions. If the server does not do this, then the rule makes no difference (and by Hyrum's law, we really shouldn't have some but not all servers rejecting non-canonical paths); it just becomes something that the client can manage. We could even imagine a "canonical enforcement proxy" which proxies requests, after verifying that they meet the canonicalization requirements.

I personally do not feel at all comfortable trying to specify canonicalization requirements for Windows. For POSIX, I *think* that we could get away with the rules about `//` and `/./`. It may be easier to explicitly frame the requirements here as guidelines, rather than hard rules for the API. And honestly, that should be considered for all canonicalization requirements unless all the implementers are enforcing them or prepared to take that on as a priority.

Reply:
The working directory does not influence the resolution of absolute symlinks. However, as the workers for BuildStream will set the filesystem root to the input root, allowing absolute paths is the same as treating them as relative to the input root. I.e., restricting the enum to just allow/disallow works for us. Let me know if I shall add this enum to the commit.

Reply:
> Let me know if I shall add this enum to the commit.

Yes, please! Even though we are effectively making it be a boolean, an enum allows us to revisit this decision later on.

Reply:
I've updated the branch based on the comments above. I've prefixed the enum values with `ABSOLUTE_SYMLINK_` to avoid conflict with `UNKNOWN` in `DigestFunction` as reported by `protoc`. Let me know if I should use a different name as prefix or if the conflict should be resolved in a different way.